go/vt/sqlparser/comments.go

```diff
@@ -266,6 +271,221 @@ func (c *ParsedComments) Directives() *CommentDirectives {
+				// If the variable matches what we are looking for, we can change its value.
+				// Otherwise we add the comment as is to our final comments and move on.
+				setVarName, _, isValid := strings.Cut(ohContent, "=")
+				if !isValid || !strings.EqualFold(strings.TrimSpace(setVarName), key) {
+					finalComment += fmt.Sprintf(" %v(%v)", ohName, ohContent)
+					continue
+				}
+				if strings.EqualFold(strings.TrimSpace(setVarName), key) {
+					keyPresent = true
+					finalComment += fmt.Sprintf(" %v(%v=%v)", ohName, strings.TrimSpace(setVarName), value)
+				}
+			} else {
+				// If it doesn't match, we add it to our final comment and move on.
+				finalComment += fmt.Sprintf(" %v(%v)", ohName, ohContent)
+			}
+		}
+		// If we haven't found any SET_VAR optimizer hint with the matching variable,
+		// then we add a new optimizer hint to introduce this variable.
+		if !keyPresent {
+			finalComment += fmt.Sprintf(" %v(%v=%v)", OptimizerHintSetVar, key, value)
+		}
+
+		finalComment += " */"
+		newComments = append(newComments, finalComment)
+	}
+	// If we have not seen even a single comment that has the optimizer hint prefix,
+	// then we add a new optimizer hint to introduce this variable.
+	if !seenFirstOhComment {
+		newComments = append(newComments, fmt.Sprintf("/*+ %v(%v=%v) */", OptimizerHintSetVar, key, value))
+	}
+	return newComments
+}
+
+// getOptimizerHint goes over the comment string from the given initial position.
+// It returns back the final position of the cursor, along with the start and end of
+// the optimizer hint name and content.
+func getOptimizerHint(initialPos int, commentStr string) (pos int, ohNameStart int, ohNameEnd int, ohContentStart int, ohContentEnd int) {
+	ohContentEnd = -1
+	// skip spaces as they aren't interesting.
+	pos = skipBlanks(initialPos, commentStr)
+	ohNameStart = pos
+	pos++
+	// All characters until we get a space of a opening bracket are part of the optimizer hint name.
+	for pos < len(commentStr) {
+		if commentStr[pos] == ' ' || commentStr[pos] == '(' {
+			break
+		}
+		pos++
+	}
+	// Mark the end of the optimizer hint name and skip spaces.
+	ohNameEnd = pos
+	pos = skipBlanks(pos, commentStr)
+	// Verify that the comment is not malformed. If it doesn't contain an opening bracket
+	// at the current position, then something is wrong.
+	if pos >= len(commentStr) || commentStr[pos] != '(' {
+		return
+	}
+	// Seeing the opening bracket, marks the start of the optimizer hint content.
+	// We skip over the comment until we see the end of the parenthesis.
+	pos++
+	ohContentStart = pos
+	pos = skipUntilParenthesisEnd(pos, commentStr)
+	ohContentEnd = pos
+	return
+}
+
+// skipUntilParenthesisEnd reads the comment string given the initial position and skips over until
+// it has seen the end of opening bracket.
+func skipUntilParenthesisEnd(pos int, commentStr string) int {
+	// This is a problem of bracket matching. We solve this using a stack, because strings are also part of the comment
```

Comment:
I think "having to use a stack" is a good hint that this shouldn't be rolled by hand and instead should be a proper YACC parser.

Reply:
I set out to change the parsing and instead use a YACC parser yesterday. While doing it I realised that the optimiser hint language doesn't actually permit nested parenthesis, and I don't really need the stack in this implementation. So I have been able to simplify this significantly. 
I can still move to a YACC parser if that is preferable, but that seems to be a lot of code, including the tokeniser, and yacc file and structs to store intermediate results.

Reply:
Happy to see this can be simplified without using YACC.